Refactor to allow configuration file

mathias opened the pull request:
Context:

We'd like to have more things be configurable in Heroics, so that the client generated can have a wider range of behaviors and functionality. (In particular, I'd like to start allowing a list of replacement patterns for the `Naming.ruby_name` method.)

The problem is, it doesn't make sense to keep tacking on more CLI args to the `heroics-generate` command.

Instead, this starts a pattern for a configuration file in Ruby. Since we don't have something like Rails to run this like an initializer, the heroics code must require these passed-in config files. The `Configuration.defaults` object sticks around to provide access to the "state" during `heroics-generate` -- although not all "state" has been migrated here, it seems to have a good delineation from the logic right now.

In the future, we will allow new, optional config values to be set in the configuration file. For now, this just replicates exactly what you can set on the existing CLI, and updates the existing CLI invocation to use the new Configuration class block style.

Other notes:
This also introduces an 'example client' in the `example/` directory, which is ignored when packaging them gem up in the gemspec. The example client can be used to see how to run the commands, how a Rake file might be wired up, and what a configuration file looks like. My other plan with including the example client was to check it in and ensure that in the future, any changes we make to heroics are verified by changes that happen when we re-run `bundle exec rake build` in the `example/` directory.

djcp commented on heroics.gemspec:
```diff
@@ -15,7 +15,7 @@ Gem::Specification.new do |spec|
   spec.license       = 'MIT'
 
-  spec.files         = `git ls-files`.split($/)
+  spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features|example)/}) }
```
TIL about  the "-z" option. Nice.

mathias commented:
I must confess that I copied this off of what is generated by default when
you run bundler gem.  It has the benefit of allowing me to be sure that the
example client doesn't make it into the built gem
 .
On Tue, Jan 31, 2017 at 1:27 PM Daniel Collis-Puro <notifications@github.com>
wrote:

> *@djcp* commented on this pull request.
> ------------------------------
>
> In heroics.gemspec
> <https://github.com/interagent/heroics/pull/80#pullrequestreview-19401673>
> :
>
> > @@ -15,7 +15,7 @@ Gem::Specification.new do |spec|
>    spec.homepage      = 'https://github.com/interagent/heroics'
>    spec.license       = 'MIT'
>
> -  spec.files         = `git ls-files`.split($/)
> +  spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features|example)/}) }
>
> TIL about the "-z" option. Nice.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/interagent/heroics/pull/80#pullrequestreview-19401673>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANDCqx7H1H423oFrhtOxQK7FIAe6fzks5rX4s5gaJpZM4Ly6Hq>
> .
>
-- 
Sent from my supercomputing distraction machine.

djcp commented on example/schema.json:
```diff
@@ -0,0 +1,245 @@
+{
```
does this example adequately cover all the naming issues we're ultimately trying to fix?

mathias commented:
It does not, but I can change it so that it does exercise thise things.
This is a copy of the schema found (as YAML) in the heroics test helper.

Note that this PR still doesn't change the ruby_name method or make it
configurable.

On Tue, Jan 31, 2017 at 2:31 PM Daniel Collis-Puro <notifications@github.com>
wrote:

> *@djcp* commented on this pull request.
> ------------------------------
>
> In example/schema.json
> <https://github.com/interagent/heroics/pull/80#pullrequestreview-19415682>
> :
>
> > @@ -0,0 +1,245 @@
> +{
>
> does this example adequately cover all the naming issues we're ultimately
> trying to fix?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/interagent/heroics/pull/80#pullrequestreview-19415682>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAANDHCQKlegQQ4j_o87N99aOMYDFD6-ks5rX5odgaJpZM4Ly6Hq>
> .
>
-- 
Sent from my supercomputing distraction machine.

djcp commented on bin/heroics-generate:
```diff
@@ -26,15 +30,36 @@ option_parser = OptionParser.new do |opts|
+case ARGV.length
+when 1
+  config_file = ARGV[0]
+  if File.readable?(config_file)
```
It might be nice to have a broader rescue somewhere to catch when a config file has syntax errors

mathias replied:
That's a good point -- I will put a test around that and then fix it. Thanks!

mathias replied:
I've actually dropped the check here, after starting to write some tests here, I'd prefer that Ruby tells us about file read or Ruby parsing errors.

djcp commented on lib/heroics/client_generator.rb:
```diff
@@ -19,8 +19,22 @@ def self.generate_client(module_name, schema, url, options)
     eruby.evaluate(context)
   end
 
+  def self.generate_client
```
did you mean to have two methods with the same name?

mathias replied:
Nope, this was a refactoring mistake.

mathias commented:
I've documented the new config file in the README. Let me know what you think!

mathias commented:
[Rendered README file changes](https://github.com/mathias/heroics/blob/5ff25d75a033bda39e355e09d053332014d4f938/README.md#configuration-file)

djcp commented on bin/heroics-generate:
```diff
@@ -26,15 +30,32 @@ option_parser = OptionParser.new do |opts|
+  require config_file
+
+  puts Heroics.generate_client
+when 3
```
I know it's obvious now, but it might be nice to throw in some comments or even an extracted function that makes it totally evident the difference between these code paths.

mathias replied:
You're right. We're walking a fine line between "This is near line-noise Ruby script from the Perl world that exists purely to invoke our OOP goodness" and "This is Ruby code that should follow our styles."

mathias replied:
Comments in cb596eeed6bcefeecbe41842a364928ec4396906. Not sure about extraction at this point, but it'd be nice if `ARGV` was less of an opaque thing that we pull magical values out of.

djcp commented on README.md:
```diff
@@ -19,6 +19,31 @@ Or install it yourself as:
 
+### Configuration File
+
+If you don't want to pass config to the CLI, you can provide a Ruby config file to the `heroics-generate` script as a single parameter.
```
Again, this is totally obvious but it might be nice to have a usage example.

mathias replied:
Yep, will add!

mathias replied:
Fixed in 25363935aeed645f9502f260065c18f609b6088e.

djcp commented:
This is great and should make it much easier to do some more .. ahem.. heroic things in the future with this.

djcp commented:
I think the last things to do here are:

1) Bump the version 0.0.18?
2) Merge
3) release the gem.

I don't know who has the release powers - @geemus and @jkakar it looks like.

mathias commented:
@geemus this doesn't modify the existing CLI invocation but allows for a config file to be used. Are we good with merging it and bumping heroics version?

geemus commented:
Sounds fine to me.

geemus commented:
Thanks!
